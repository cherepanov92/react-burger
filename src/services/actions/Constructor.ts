import { OrderedIngredient } from '../../utils/types';

export const ADD_INGREDIENT: 'ADD_INGREDIENT' = 'ADD_INGREDIENT';
export const REMOVE_INGREDIENT: 'REMOVE_INGREDIENT' = 'REMOVE_INGREDIENT';
export const MOVE_INGREDIENT: 'MOVE_INGREDIENT' = 'MOVE_INGREDIENT';

interface IAddIngredient {
    readonly type: typeof ADD_INGREDIENT;
    ingredient: OrderedIngredient;
}

interface IRemoveIngredient {
    readonly type: typeof REMOVE_INGREDIENT;
    ingredient: OrderedIngredient;
}

interface IMoveIngredient {
    readonly type: typeof MOVE_INGREDIENT;
    ingredient: OrderedIngredient;
    newIndex: number;
}

export type TIngredientActions = IAddIngredient | IRemoveIngredient | IMoveIngredient;