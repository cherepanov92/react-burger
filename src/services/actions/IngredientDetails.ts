import { IngredientType } from '../../utils/types';

export const ADD_INGREDIENT_DETAILS: 'ADD_INGREDIENT_DETAILS' = 'ADD_INGREDIENT_DETAILS';
export const REMOVE_INGREDIENT_DETAILS: 'REMOVE_INGREDIENT_DETAILS' = 'REMOVE_INGREDIENT_DETAILS';

interface IAddIngredientDetails {
    readonly type: typeof ADD_INGREDIENT_DETAILS;
    ingredient: IngredientType;
}

interface IRemoveIngredientDetails {
    readonly type: typeof REMOVE_INGREDIENT_DETAILS;
}

export type TIngredientDetailsActions = IAddIngredientDetails | IRemoveIngredientDetails;