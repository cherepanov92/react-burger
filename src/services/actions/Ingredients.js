import { getIngredients } from "../../utils/api";
import { APPEND_ERROR_MODAL_TYPE } from "./Modal";

export const GET_INGREDIENTS_REQUEST = 'GET_INGREDIENTS_REQUEST';
export const GET_INGREDIENTS_SUCCESS = 'GET_INGREDIENTS_SUCCESS';
export const GET_INGREDIENTS_FAILED = 'GET_INGREDIENTS_FAILED';

export const getIngredientsData = () => {
    return function(dispatch) {
        dispatch({
            type: GET_INGREDIENTS_REQUEST
        });
        getIngredients()
        .then(res => {
            if (res && res.success) {
                dispatch({
                    type: GET_INGREDIENTS_SUCCESS,
                    items: res.data
                });
            } else {
                dispatch({
                    type: GET_INGREDIENTS_FAILED
                });
            }
        })
        .catch(err => {
            dispatch({
                type: APPEND_ERROR_MODAL_TYPE,
                errorStatus: err.status
            })
        })
    };
}