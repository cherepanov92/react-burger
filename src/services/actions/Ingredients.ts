import { getIngredients } from '../../utils/api';
import { APPEND_ERROR_MODAL_TYPE } from './Modal';
import { IError, IngredientType } from '../../utils/types';
import { Dispatch } from 'react';

export const GET_INGREDIENTS_REQUEST: 'GET_INGREDIENTS_REQUEST' = 'GET_INGREDIENTS_REQUEST';
export const GET_INGREDIENTS_SUCCESS: 'GET_INGREDIENTS_SUCCESS' = 'GET_INGREDIENTS_SUCCESS';
export const GET_INGREDIENTS_FAILED: 'GET_INGREDIENTS_FAILED' = 'GET_INGREDIENTS_FAILED';

interface IGetIngredientsRequest {
    readonly type: typeof GET_INGREDIENTS_REQUEST;
}

interface IGetIngredientsSuccess {
    readonly type: typeof GET_INGREDIENTS_SUCCESS;
    items: IngredientType[];
}

interface IGetIngredientsFailed {
    readonly type: typeof GET_INGREDIENTS_FAILED;
}

export type TIngredientsActions = IGetIngredientsRequest | IGetIngredientsSuccess | IGetIngredientsFailed;

export const getIngredientsData = () => {
    return function (dispatch: Dispatch<TIngredientsActions | IError>) {
        dispatch({
            type: GET_INGREDIENTS_REQUEST
        });
        getIngredients()
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: GET_INGREDIENTS_SUCCESS,
                        items: res.data
                    });
                } else {
                    dispatch({
                        type: GET_INGREDIENTS_FAILED
                    });
                }
            })
            .catch(err => {
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                });
            });
    };
};
