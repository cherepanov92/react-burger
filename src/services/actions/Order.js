import { sendOrder } from '../../utils/api';
import { APPEND_ERROR_MODAL_TYPE, APPEND_MODAL_TYPE } from './Modal';

export const SENT_ORDER_REQUEST = 'SENT_ORDER_REQUEST';
export const SENT_ORDER_SUCCESS = 'SENT_ORDER_SUCCESS';
export const SENT_ORDER_FAILED = 'SENT_ORDER_FAILED';

export const sendOrderRequest = orderList => {
    return function (dispatch) {
        dispatch({
            type: SENT_ORDER_REQUEST
        });
        sendOrder(orderList)
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: SENT_ORDER_SUCCESS,
                        name: res.name,
                        number: res.order.number
                    });
                    dispatch({ type: APPEND_MODAL_TYPE, modalType: 'order' });
                } else {
                    dispatch({
                        type: SENT_ORDER_FAILED
                    });
                }
            })
            .catch(err => {
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                });
            });
    };
};
