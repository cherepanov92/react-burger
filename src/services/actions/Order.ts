import { sendOrder } from '../../utils/api';
import { APPEND_ERROR_MODAL_TYPE, APPEND_MODAL_TYPE, IAppendModalType } from './Modal';
import { EnumModalType, IError, OrderedIngredient } from '../../utils/types';
import { Dispatch } from 'react';

export const SENT_ORDER_REQUEST: 'SENT_ORDER_REQUEST' = 'SENT_ORDER_REQUEST';
export const SENT_ORDER_SUCCESS: 'SENT_ORDER_SUCCESS' = 'SENT_ORDER_SUCCESS';
export const SENT_ORDER_FAILED: 'SENT_ORDER_FAILED' = 'SENT_ORDER_FAILED';

interface ISendOrderRequest {
    readonly type: typeof SENT_ORDER_REQUEST;
}

interface ISendOrderSuccess {
    readonly type: typeof SENT_ORDER_SUCCESS;
    name: string;
    number: number;
}

interface ISendOrderFailed {
    readonly type: typeof SENT_ORDER_FAILED;
}

export type TOrderActions = ISendOrderRequest | ISendOrderSuccess | ISendOrderFailed;

export const sendOrderRequest = (orderList: OrderedIngredient[]) => {
    return function (dispatch: Dispatch<TOrderActions | IAppendModalType | IError>) {
        dispatch({
            type: SENT_ORDER_REQUEST
        });
        sendOrder(orderList)
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: SENT_ORDER_SUCCESS,
                        name: res.name,
                        number: res.order.number
                    });
                    dispatch({ type: APPEND_MODAL_TYPE, modalType: EnumModalType.ORDER });
                } else {
                    dispatch({
                        type: SENT_ORDER_FAILED
                    });
                }
            })
            .catch(err => {
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                });
            });
    };
};
