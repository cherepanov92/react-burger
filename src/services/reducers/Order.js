import { SENT_ORDER_FAILED, SENT_ORDER_REQUEST, SENT_ORDER_SUCCESS } from "../actions/Order";

export default function OrderReducer(state = {}, action) {
    switch (action.type) {
        case SENT_ORDER_REQUEST: {
            return {
                ...state,
                orderRequest: true,
                orderFailed: false,
                orderData: { ...state.order, name: null, number: null }
            };
        }
        case SENT_ORDER_SUCCESS: {
            return {
                ...state,
                orderRequest: false,
                orderData: { ...state.order, name: action.name, number: action.number }
            };
        }
        case SENT_ORDER_FAILED: {
            return {
                ...state,
                orderFailed: true,
                orderRequest: false,
                orderData: { ...state.order, name: null, number: null }
            };
        }
        default: {
            return state
        }
    }
}