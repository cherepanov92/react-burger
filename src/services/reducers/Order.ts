import { SENT_ORDER_FAILED, SENT_ORDER_REQUEST, SENT_ORDER_SUCCESS } from '../actions/Order';

type orderState = {
    orderRequest?: boolean;
    orderFailed?: boolean;
    //todo: узнать какой тип
    order?: any;
    orderData?: {
        name: string;
        number: number;
    };
};

export default function orderReducer(state: orderState = {}, action: any) {
    switch (action.type) {
        case SENT_ORDER_REQUEST: {
            return {
                ...state,
                orderRequest: true,
                orderFailed: false,
                orderData: { ...state.order, name: null, number: null }
            };
        }
        case SENT_ORDER_SUCCESS: {
            return {
                ...state,
                orderRequest: false,
                orderData: { ...state.order, name: action.name, number: action.number }
            };
        }
        case SENT_ORDER_FAILED: {
            return {
                ...state,
                orderFailed: true,
                orderRequest: false,
                orderData: { ...state.order, name: null, number: null }
            };
        }
        default: {
            return state;
        }
    }
}
