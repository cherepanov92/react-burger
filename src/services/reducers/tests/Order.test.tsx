import React from 'react';
import renderer from 'react-test-renderer';
import orderReducer, { InitialOrderState as initialState } from '../Order';
import { SENT_ORDER_FAILED, SENT_ORDER_REQUEST, SENT_ORDER_SUCCESS } from '../../actions/Order';

describe('OrderReducer:', () => {
    it('Начальное состояние', () => {
        expect(orderReducer(initialState, {} as any)).toEqual(initialState);
        const tree = renderer.create(<>Начальное состояние</>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Запрос на создание заказа', () => {
        expect(
            orderReducer(initialState, {
                type: SENT_ORDER_REQUEST
            })
        ).toEqual({
            ...initialState,
            orderRequest: true,
            orderFailed: false,
            orderData: { name: null, number: null }
        });
    });

    it('Подтверждение заказа', () => {
        expect(
            orderReducer(initialState, {
                type: SENT_ORDER_SUCCESS,
                name: 'super_order',
                number: 1337
            })
        ).toEqual({
            ...initialState,
            orderRequest: false,
            orderData: { name: 'super_order', number: 1337 }
        });
    });

    it('Ошибка отправки заказа', () => {
        expect(
            orderReducer(initialState, {
                type: SENT_ORDER_FAILED
            })
        ).toEqual({
            ...initialState,
            orderFailed: true,
            orderRequest: false,
            orderData: { name: null, number: null }
        });
    });
});
