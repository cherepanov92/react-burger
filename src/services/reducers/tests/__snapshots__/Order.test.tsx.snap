// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderReducer: Начальное состояние 1`] = `"Начальное состояние"`;
